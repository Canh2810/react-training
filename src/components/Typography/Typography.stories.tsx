import { StoryObj, Meta } from '@storybook/react'

import Typography from './'
import { TypoVariants } from '@types'

export default {
  title: 'Components/Typography',
  component: Typography,
} as Meta

type Story = StoryObj<typeof Typography>

export const Default: Story = {
  args: {
    children: 'This is text',
  },
}

export const TextExtraSmall: Story = {
  args: {
    children: 'This is text',
    variant: TypoVariants.TextExtraSmall,
  },
}

export const TextSmall: Story = {
  args: {
    children: 'This is text',
    variant: TypoVariants.TextSmall,
  },
}

export const TextMedium: Story = {
  args: {
    children: 'This is text',
    variant: TypoVariants.TextMedium,
  },
}

export const TextLarge: Story = {
  args: {
    children: 'This is text',
    variant: TypoVariants.TextLarge,
  },
}

export const TexExtraLarge: Story = {
  args: {
    children: 'This is text',
    variant: TypoVariants.TexExtraLarge,
  },
}

export const HeadingSmall: Story = {
  args: {
    children: 'This is text',
    variant: TypoVariants.HeadingSmall,
  },
}

export const HeadingMedium: Story = {
  args: {
    children: 'This is text',
    variant: TypoVariants.HeadingMedium,
  },
}

export const HeadingLarge: Story = {
  args: {
    children: 'This is text',
    variant: TypoVariants.HeadingLarge,
  },
}

export const HeadingExtraLarge: Story = {
  args: {
    children: 'This is text',
    variant: TypoVariants.HeadingExtraLarge,
  },
}
