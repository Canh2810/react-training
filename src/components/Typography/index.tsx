// Libs
import { memo } from 'react'

// Types
import { TypoVariants, TypoColors } from '@types'

// Stores
import { useThemeStore } from '@stores'

// Themes
import { palette, variants } from '@themes'

export interface TypographyProps {
  variant?: TypoVariants
  color?: TypoColors
  children: React.ReactNode
  component?: keyof JSX.IntrinsicElements
  className?: string
}

const Typography = ({
  variant = TypoVariants.TextSmall,
  color = TypoColors.Primary,
  children,
  component = 'p',
  className = '',
}: TypographyProps) => {
  const { theme } = useThemeStore((state) => state)
  const Component = component as keyof JSX.IntrinsicElements
  const typographyStyles = {
    ...variants[variant],
    ...palette[color](theme),
  }

  return (
    <Component className={className} style={typographyStyles}>
      {children}
    </Component>
  )
}

export default memo(Typography)
