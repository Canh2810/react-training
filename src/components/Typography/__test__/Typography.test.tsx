// PostList.test.js (or PostList.test.tsx for TypeScript)

import { render } from '@testing-library/react'
import Typography from '..'
import { TypoColors, TypoVariants } from '@types'
import { useThemeStore } from '@stores'
import { THEMES } from '@constants'

describe('PostList component', () => {
  const props = {
    children: 'This is text',
  }

  beforeEach(() => {
    useThemeStore.setState({
      theme: THEMES.LIGHT,
    })
  })

  it('render with default props', () => {
    const { container } = render(<Typography {...props} />)
    expect(container).toMatchSnapshot()
  })

  it('render with variant is textExtraSmall', () => {
    const { container } = render(
      <Typography {...props} variant={TypoVariants.TextExtraSmall} />,
    )
    expect(container).toMatchSnapshot()
  })

  it('render with variant is textMedium', () => {
    const { container } = render(
      <Typography {...props} variant={TypoVariants.TextMedium} />,
    )
    expect(container).toMatchSnapshot()
  })

  it('render with variant is textLarge', () => {
    const { container } = render(
      <Typography {...props} variant={TypoVariants.TextLarge} />,
    )
    expect(container).toMatchSnapshot()
  })

  it('render with variant is textExtraLarge', () => {
    const { container } = render(
      <Typography {...props} variant={TypoVariants.TexExtraLarge} />,
    )
    expect(container).toMatchSnapshot()
  })

  it('render with variant is headingSmall', () => {
    const { container } = render(
      <Typography {...props} variant={TypoVariants.HeadingSmall} />,
    )
    expect(container).toMatchSnapshot()
  })

  it('render with variant is headingMedium', () => {
    const { container } = render(
      <Typography {...props} variant={TypoVariants.HeadingMedium} />,
    )
    expect(container).toMatchSnapshot()
  })

  it('render with variant is headingLarge', () => {
    const { container } = render(
      <Typography {...props} variant={TypoVariants.HeadingLarge} />,
    )
    expect(container).toMatchSnapshot()
  })

  it('render with variant is headingExtraLarge', () => {
    const { container } = render(
      <Typography {...props} variant={TypoVariants.HeadingExtraLarge} />,
    )
    expect(container).toMatchSnapshot()
  })

  it('render with color is textSecondary', () => {
    const { container } = render(
      <Typography {...props} color={TypoColors.Secondary} />,
    )
    expect(container).toMatchSnapshot()
  })

  it('render with color is textTertiary', () => {
    const { container } = render(
      <Typography {...props} color={TypoColors.Tertiary} />,
    )
    expect(container).toMatchSnapshot()
  })

  it('render with component is h1', () => {
    const { container } = render(<Typography {...props} component="h1" />)
    expect(container).toMatchSnapshot()
  })
})
