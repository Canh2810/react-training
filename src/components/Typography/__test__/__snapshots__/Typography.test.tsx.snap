// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostList component render with color is textSecondary 1`] = `
<div>
  <p
    class=""
    style="font-family: plus-jakarta-sans; font-size: 16px; font-weight: 500; line-height: 24px; color: rgb(24, 26, 42);"
  >
    This is text
  </p>
</div>
`;

exports[`PostList component render with color is textTertiary 1`] = `
<div>
  <p
    class=""
    style="font-family: plus-jakarta-sans; font-size: 16px; font-weight: 500; line-height: 24px; color: rgb(105, 106, 117);"
  >
    This is text
  </p>
</div>
`;

exports[`PostList component render with component is h1 1`] = `
<div>
  <h1
    class=""
    style="font-family: plus-jakarta-sans; font-size: 16px; font-weight: 500; line-height: 24px; color: rgb(59, 60, 74);"
  >
    This is text
  </h1>
</div>
`;

exports[`PostList component render with default props 1`] = `
<div>
  <p
    class=""
    style="font-family: plus-jakarta-sans; font-size: 16px; font-weight: 500; line-height: 24px; color: rgb(59, 60, 74);"
  >
    This is text
  </p>
</div>
`;

exports[`PostList component render with variant is headingExtraLarge 1`] = `
<div>
  <p
    class=""
    style="font-family: work-sans; font-size: 36px; font-weight: 600; line-height: 40px; color: rgb(59, 60, 74);"
  >
    This is text
  </p>
</div>
`;

exports[`PostList component render with variant is headingLarge 1`] = `
<div>
  <p
    class=""
    style="font-family: work-sans; font-size: 24px; font-weight: 600; line-height: 28px; color: rgb(59, 60, 74);"
  >
    This is text
  </p>
</div>
`;

exports[`PostList component render with variant is headingMedium 1`] = `
<div>
  <p
    class=""
    style="font-family: work-sans; font-size: 20px; font-weight: 600; line-height: 28px; color: rgb(59, 60, 74);"
  >
    This is text
  </p>
</div>
`;

exports[`PostList component render with variant is headingSmall 1`] = `
<div>
  <p
    class=""
    style="font-family: plus-jakarta-sans; font-size: 18px; font-weight: 600; line-height: 28px; color: rgb(59, 60, 74);"
  >
    This is text
  </p>
</div>
`;

exports[`PostList component render with variant is textExtraLarge 1`] = `
<div>
  <p
    class=""
    style="font-family: source-serif-pro; font-size: 24px; font-weight: 400; line-height: 32px; color: rgb(59, 60, 74);"
  >
    This is text
  </p>
</div>
`;

exports[`PostList component render with variant is textExtraSmall 1`] = `
<div>
  <p
    class=""
    style="font-family: inter; font-size: 14px; font-weight: 400; line-height: 20px; color: rgb(59, 60, 74);"
  >
    This is text
  </p>
</div>
`;

exports[`PostList component render with variant is textLarge 1`] = `
<div>
  <p
    class=""
    style="font-family: source-serif-pro; font-size: 20px; font-weight: 400; line-height: 32px; color: rgb(59, 60, 74);"
  >
    This is text
  </p>
</div>
`;

exports[`PostList component render with variant is textMedium 1`] = `
<div>
  <p
    class=""
    style="font-family: work-sans; font-size: 18px; font-weight: 400; line-height: 26px; color: rgb(59, 60, 74);"
  >
    This is text
  </p>
</div>
`;
